packages/drivers/odsp-driver/src/opsCaching.ts

```diff
@@ -0,0 +1,170 @@
+
+export interface IBatch {
+    remainingSlots: number;
+    batchData: CacheEntry;
+    /**
+     * Tells if this batch is  dirty, i.e. it contains ops that were not flushed to cache
+     */
+    dirty: boolean;
+}
+
+export interface ICache {
+  write(batchNumber: string, data: string): Promise<void>;
+  read(batchNumber: string): Promise<string | undefined>;
+}
+
+export class OpsCache {
+  private readonly batches: Map<number, null | IBatch> = new Map();
+  private timer: ReturnType<typeof setTimeout> | undefined;
+
+  constructor(
+    startingSequenceNumber: number,
+    private readonly cache: ICache,
+    private readonly batchSize: number,
+    private readonly timerGranularity,
+    private totalOpsToCache,
+  ) {
+        /** initial batch is a special case because it will never be full - all ops prior (inclusive) to
+         *  startingSequenceNumber are never going to show up (undefined)
+         */
+        const remainingSlots = this.batchSize - this.getPositionInBatchArray(startingSequenceNumber) - 1;
+        if (remainingSlots !== 0) {
+            this.batches.set(this.getBatchNumber(startingSequenceNumber), {
+                remainingSlots,
+                batchData : this.initializeNewBatchDataArray(),
+                dirty: false,
+            });
+        }
+    }
+
+    public flushOps() {
+        for (const [key, value] of this.batches) {
+            if (value === null || !value.dirty) {
+                continue;
+            }
+            value.dirty = false;
+            this.write(key, value);
+        }
+    }
+
+    public addOps(ops: IMessage[]) {
+        if (this.totalOpsToCache <= 0) {
+            return;
+        }
+
+        for (const op of ops) {
+            const batchNumber = this.getBatchNumber(op.sequenceNumber);
+            const positionInBatch = this.getPositionInBatchArray(op.sequenceNumber);
+
+            let currentBatch = this.batches.get(batchNumber);
+
+            if (currentBatch === undefined) {
+                currentBatch = {
+                    remainingSlots: this.batchSize - 1,
+                    batchData: this.initializeNewBatchDataArray(),
+                    dirty: true,
+                };
+                currentBatch.batchData[positionInBatch] = op;
+                this.batches.set(batchNumber, currentBatch);
+            } else if (currentBatch !== null && currentBatch.batchData[positionInBatch] === undefined) {
+                currentBatch.batchData[positionInBatch] = op;
+                currentBatch.remainingSlots--;
+                currentBatch.dirty = true;
+            } else {
+                // Either batch was flushed or this op was already there - nothing to do!
+                return;
+            }
+
+            if (currentBatch.remainingSlots === 0) {
+                // batch is full, flush to cache
+                this.write(batchNumber, currentBatch);
+                this.batches.set(batchNumber, null);
```

Comment:
So here, we are flushing this batch to cache but not setting the dirty flag to false.

Reply:
That's correct, there is no need for it, as it will get deleted (setting null) from the list that is more efficient way of storing same info - it indicates that this entry no longer wants any ops as it was full. Keeping object itself (and setting dirty = false) would work as well, but it would take way more memory than having an array of nulls.

What is missing is UT case to validate ops that come to that batch will not flush it any more, I'll add it.